Indicator Heterogeneity I

huisaddison opened the pull request:
This PR signals the formal submission of the DAP Report for Indicator Heterogeneity I and closes #18.

Read the report here: https://htmlpreview.github.io/?https://github.com/cmu-delphi/covidcast-modeling/blob/devel/indicator-heterogeneity-I/indicator_heterogeneity_I/reports/DAP_Report_IndicatorHeterogeneity.html

> In this notebook, we perform a deep investigation of one particular
> sensorization procedure: a moving window regression, where we fit
> using data from 10 weeks back to 1 week back.  **We find that dynamic
> sensorization is able to correct spatial heterogeneity (improve geo-wise
> correlations), with neutral effects on time-wise correlations and
> ability to forecast future cases/deaths.**  Validation setups suggest
> that it _may_ be possible that we are overfitting to our targets.
> I would be far more confident geo-wise correlation results if we get some
> measures of spread (e.g., bootstrap confidence intervals).

ryantibs commented on indicator_heterogeneity_I/reports/DAP_Report_IndicatorHeterogeneity.Rmd:
```diff
@@ -0,0 +1,881 @@
+   expect that order(hospitalizations) ≈ order(cases), this gives evidence that
+   sensorization _is_ correcting spatial heterogeneity in the Doctor visits
+   indicator!
+3. It is possible that (1) is because there is not that much spatial
+   heterogeneity in the hospitalizations indicator to begin with.
```
Yes that might be right.  But then something a bit disturbing is happening in the geo-wise correlations above: we see substantial improvement when sensorizing the hospitalization indicator.  This must just be because we're leaking info from the deaths into the sensorized indicator?  (Overfitting?)  Reveals the fundamental concern we have with sensorization all along, is that we could be discarding info from the indicator and entering info from the response, basically reducing independent information and increasing duplication.

huisaddison replied:
That is true.  I will also comment that Hospitalizations sees improvement when sensorized against Cases, but not when being sensorized against Deaths.  Another cautionary note that Cases is "heavy handed" in inserting its information into the sensorized indicator?

ryantibs commented on indicator_heterogeneity_I/reports/DAP_Report_IndicatorHeterogeneity.Rmd:
```diff
@@ -0,0 +1,881 @@
+"improvements" were merely the by product of replicating cases in the
+intercept.
+
+**The coefficient distributions leave much to be desired.**  In the perfect
+world of our "data model", the intercepts would be zero (e.g., no COVID-related
+doctor visits would correspond to no cases); and the slopes would carry the
+information, perhaps varying over time to account for temporal heterogeneity.
+
+What we see in our data is quite different.  The intercepts tend to be away
+from zero, while the slopes tend to be near zero (often, the median ± mad
+interval covers zero), despite our purposeful choice of a wide, nine-week
+wide training window to avoid trivially fitting a moving average.
```
Hmm well-said, and too bad.  But the slopes aren't as bad as I've seen them in your (or Aaron's? I forget) previous analysis.  They still seem decently large here.

ryantibs commented on indicator_heterogeneity_I/reports/DAP_Report_IndicatorHeterogeneity.Rmd:
```diff
@@ -0,0 +1,881 @@
+the Doctor visits, Facebook %CLI, and Hospitalizations signals when
+sensorized against Cases.
+
+Independent of the "meaning" of this simultaneous increase in the
+intercepts and decrease in the slopes, it is especially concerning that
+it happens for (nearly) all the indicators sensorized against Cases, at
+the same time.  This suggests that the shifts are due to a change in
+the behavior of Cases, as a simultaneous change in the behavior of
+the indicators is unlikely.  We should be especially mindful of this,
+i.e., that the "heterogeneity" being picked up by sensorization can be
+instability in the target data.
```
Great point.

ryantibs reviewed (commented):
Overall, this is absolutely awesome.  Nice job!  

Hope addressing my comments doesn't entail too much more work (don't want to drag this out much more, just want to close the loop on the things I raised).

jacobbien commented on indicator_heterogeneity_I/reports/DAP_Report_IndicatorHeterogeneity.Rmd:
```diff
@@ -0,0 +1,881 @@
+We fit a linear model for each location and day, and then take the prediction
+for that day.
+
+For our purposes we take $k_1, k_2$ to be $-70, -8$.  That is, we train over
```
If it's $x_{(t-k_1):(t-k_2), \ell}$, then should $k_1$ and $k_2$ be positive rather than negative?

huisaddison replied:
Yes - I have fixed this.  Thanks!

huisaddison commented on indicator_heterogeneity_I/reports/DAP_Report_IndicatorHeterogeneity.Rmd:
```diff
@@ -0,0 +1,881 @@
+
+We fit, for each $t, \ell$:
+$$
+y_{(t-k):(t-1), \ell} \sim x_{(t-k_1):(t-k_2), \ell} \qquad\Rightarrow
```
Yes, that's right.  I have fixed this.

huisaddison commented on indicator_heterogeneity_I/reports/DAP_Report_IndicatorHeterogeneity.Rmd:
```diff
@@ -0,0 +1,881 @@
+indicator's values are (more) comparable across location.  This is important,
+for example, if we want the indicator to be intuitively useful on a map.
+
+**We generally find that the geo-wise correlations are improved by
+sensorization, both dynamic and static.**  This is encouraging.  It is
+noteworthy that dynamic sensorization does essentially as well as static
+sensorization here, since this static sensorization has the added advantage
+that estimates are being produced in-sample.
```
(1) I will rerun the sensorization and update the notebook.

(2) Yes, I am using smoothed cases and deaths.  I have made this explicit in the text.

huisaddison commented on indicator_heterogeneity_I/reports/DAP_Report_IndicatorHeterogeneity.Rmd:
```diff
@@ -0,0 +1,881 @@
+component out of this data by taking the median (and mad) of the correlations
+for each day.
+
+Previously, Aaron found (communicated on Slack) that dynamic sensorization
+appeared to degrade the time-wise correlation of Doctor visits against cases.
+I reproduced this apparent result, but also plotted the spreads of the
+time-wise correlations.  The distributions of the time-wise correlations
+are overlapping, so I am not as concerned.  I believe that the most we
+can say is that dynamic sensorization does not provide any added advantage
+in time-wise correlations, compared to static sensorization or the raw
+indicator.
```
(1) Yes, the correlations here are on the same scale as what Aaron previously found: https://delphi-org.slack.com/archives/C011EU72MU2/p1606510990040900?thread_ts=1606316883.034700&cid=C011EU72MU2

(2) Red vs green is a good point, and I agree with you on principle.  I don't know why they are different yet, but I will keep this thread open and follow up once I have more detail.s

huisaddison replied:
For (2): when we look by location, the time-wise cors for the raw and static sensors are identical.  The discrepancy in the median/mad come in the fact that in a limited number of cases, a static sensorization is not available (e.g., due to missing data; observe that the Hospitalizations indicator suffers more from this problem / some kind of error in the sensorization step).  

Since the difference is artificial, I can omit the static sensorization timewise cors in the final plots, and note that they are identical to the raw timewise cors.

huisaddison commented on indicator_heterogeneity_I/reports/DAP_Report_IndicatorHeterogeneity.Rmd:
```diff
@@ -0,0 +1,881 @@
+be large across locations, but perhaps the dynamic sensorization fares
+consistently worse than the raw indicator by a small amount.
+
+What we find is more peculiar.  We find that the median pointwise differences
+between time-wise rank correlations of the sensorized and the raw
+indicator actually are positive.  But again, the spreads are large.
```
Ok, I will do this for the updated report.

huisaddison replied:
Figured it out - it was a bug not a paradox (bug was that I forgot that assignment takes place 'right away' inside mutate()): 
https://github.com/cmu-delphi/covidcast-modeling/pull/22/files#diff-4328b296cee17e5556717b5ef9e312d3d4e09bf83b137eb57b5d3e5c22fd3498R380-R382

Once fixed, the plot is much more sane and basically what we expect:

![Screenshot from 2020-12-15 22-53-57](https://user-images.githubusercontent.com/8422804/102315433-00872d00-3f29-11eb-83c1-bc1aafe8fa54.png)

huisaddison commented on indicator_heterogeneity_I/reports/DAP_Report_IndicatorHeterogeneity.Rmd:
```diff
@@ -0,0 +1,881 @@
+
+### Predictive ability
+
+For each (indicator, target) pair, we perform the static and dynamic
+sensorizations, and then evaluate its ability to "improve upon" the
+predictive ability of the target.  The setup is the same as in Ryan's
+forecasting notebook.  We perform 7 day ahead and 14 day ahead forecasts.
+The base model only uses the target information (cases or deaths),
+at lags of 7 days behind and 14 days behind.  The other three models
+add the raw indicator, static sensorized indicator, and dynamic sensorized
+indicator to the target (all three at the corresponding lags).
```
Yes - three features in each (I forgot that `lag = 0` was in the model because the _appended_ `lags` were `1:2 * -7`.  I will update this.

huisaddison commented on indicator_heterogeneity_I/reports/DAP_Report_IndicatorHeterogeneity.Rmd:
```diff
@@ -0,0 +1,881 @@
+strawman, which propogates the previous week forward.  Scaled errors
+below 1 are desirable.
+
+**We find, on the whole, that including the sensorized indicator neither helps
+nor hurts in the prediction task.**  This is discouraging for us; we might hope
+that sensorization, when performed properly, would provide "additional
+information" beyond the target.
```
(1) Yes, median -- the legend states this by describing the linetype, but I will make it clear in the text.

(2) Sure, I will do this for the final report.

(3) I meant it's discouraging that the models with the indicators (sensorized or not) do essentially the same as the models with no indicators (just cases or just deaths, i.e., just the (lagged) target).  But yes, they all do better than the strawman.

(4) Yeah, in principle we should see gains from putting the different counties on the same scale...

(5) Sure, I will rerun this part as well.

huisaddison commented on indicator_heterogeneity_I/reports/DAP_Report_IndicatorHeterogeneity.Rmd:
```diff
@@ -0,0 +1,881 @@
+
+### Geo-wise correlations
+
+For each of our sensorized indicators, we compute its geo-wise rank correlation
+against the target data.  Assuming that the target data observes the proper
+"ordering" for each day, an improvement in the geo-wise rank correlations
+in the sensorized indicators would mean that for a fixed day, the (sensorized)
+indicator's values are (more) comparable across location.  This is important,
+for example, if we want the indicator to be intuitively useful on a map.
```
I am considering all counties with at least 500 cumulative cases by November 1st, 2020.  (And all US States).  Let me know if I should increase the threshold when re-running under the new settings.

huisaddison commented:
@ryantibs @jacobbien my latest commit updates the DAP report: https://htmlpreview.github.io/?https://github.com/cmu-delphi/covidcast-modeling/blob/devel/indicator-heterogeneity-I/indicator_heterogeneity_I/reports/DAP_Report_IndicatorHeterogeneity.html

Changes: 

* Fixed typos / clarified text
* Reran setup such that sensorized estimates are available starting April 1st (specifically, we ingest raw data starting from March 15th, 2020, and we require a week of data to start sensorizing)
* (Note that the hospitalization validation still starts mid-July, since that's when HHS starts providing data for nearly all states)
* Removed +/- mad intervals from the predictive exercise plots, to improve clarity
* Fixed a bug in one of the timewise correlation plots.  The "peculiarity" is no longer there.

ryantibs commented:
Great job Addison, that last revision cleared a lot of things up.

One last point sticks in my mind: in the [predictive ability section](https://htmlpreview.github.io/?https://github.com/cmu-delphi/covidcast-modeling/blob/devel/indicator-heterogeneity-I/indicator_heterogeneity_I/reports/DAP_Report_IndicatorHeterogeneity.html#predictive-ability), you basically conclude that all models perform equally well for each task, but that's a bit to blunt of a conclusion as far as I see it?  Sure, the median scaled errors all look similar (and the median absolute deviations were really big, I know you committed them), but don't think this means everything is faring the same way.  

In my blog post I found pretty similar looking results (for the CLI-in-community signal, over a shorter period of time), and based on the simple inferences I performed there (admittedly, not totally rigorous due to a lack of independence), the results looked significant.  Most compelling to me (intuitively, not rigorously) was actually the part where I [varied the number of days[(https://delphi.cmu.edu/blog/2020/09/21/can-symptoms-surveys-improve-covid-19-forecasts/#varying-the-number-of-days-ahead) ahead and saw how the errors behaved as a function of the forecast horizon.  

So if I'm reading your results kind of off-the-cuff, I see (in each case comparing to the base model):

- For DV, adding the static DV sensor actually helps quite a bit, raw DV sensor somewhat, and disappointingly (!) it's unclear for the dynamic DV sensor.

- Same story for CLI and CLI-in-community.

- Same story for hospitalizations with deaths as the target, and the differences look even bigger here.

Do you agree?  It's quite disappointing that the dynamic sensorization actually "ruins" what you see with the raw sensor.  We'd obviously hope it'd go the other way, where it improves on it (and gets close to static sensorization).  Particularly for DV where we know it's a big problem.  Any thoughts on this?

How much work would it be for you to repeat the prediction task over multiple aheads (like 5 through 20, as I did in my forecast demo), and produce plots of error versus ahead?  That might give us some more insight into what's going on?

Before you start any work in earnest, let me hear your reaction, and estimate for me how much time the above would take (rerun over varying aheads).  I really don't want to drag this out---you've done a terrific job and we've gotten some really useful insights already.  But to the extent we can close the last mile here and see what's happening with prediction error, I think it'd really help inform modeling.

huisaddison commented:
Hey @ryantibs -- your conclusion that in the predictive task, static does better than dynamic makes me think that you are reading the second of the two "predictive ability" plots:

* The first predictive ability plot has the base covariates (14, 7, 0 days lag) and the sensorized covariates (14, 7, 0). When I read this plot, static edges out dynamic for DV, dynamic does slightly better than static for Facebook CLI and CLI-in-Community, and the results for Hospitalizations is mixed
<img width="1164" alt="Screen Shot 2020-12-16 at 22 18 41" src="https://user-images.githubusercontent.com/8422804/102451075-ab145400-3fec-11eb-838e-bdd945df7e4d.png">

* The second predictive ability plot only has the sensorized covariates (no base covariates).  In this set of plots, static does consistently better than dynamic

<img width="1172" alt="Screen Shot 2020-12-16 at 22 18 59" src="https://user-images.githubusercontent.com/8422804/102451092-b4052580-3fec-11eb-9b3e-31bd7cc54ab7.png">

Were you looking at the latter when you made those conclusions? (And if so, sorry for the confusion - I made the latter plot as a "sanity check" / "validation" since the lines in the former plot were so close to each other).

Regarding the extra aheads - it's a simple code change, so I'm already running them overnight.  And then we can take a quick peek of the results tomorrow, to see if it reveals anything more that we don't know yet.

ryantibs commented:
I was actually looked at the table (overall medians, across all time).

huisaddison commented:
Ah I see.  Yes then I totally agree.  The was a slight hiccup in the multiple leads script but the new results should be available tonight (don't worry, it's not taking too much working time, just running it in the background).

ryantibs commented:
Good to know.  Btw: beware of over what dates you aggregate the results when drawing the error curves as a function of lead time!  I just did reran my own script this morning, to update a talk I was giving, and was surprised to find the answers looking totally different pre the third wave, and during the third wave.  I think I understand in hindsight what's going on now.  But good to be aware of it.  

Check out: https://cmu-delphi.github.io/covidcast/talks/copss-niss/talk.html#(34) versus https://cmu-delphi.github.io/covidcast/talks/copss-niss/talk.html#(35)

huisaddison commented:
Oh, wow, good to know! That fits into what we know about the strawman's performance during and not during the surge, but the relative performance of the different models complicates the story.

I'll make sure to facet on a couple different time ranges in my final plots.

huisaddison commented:
@ryantibs The report has been updated.  Link to the [predictive ability section](https://htmlpreview.github.io/?https://github.com/cmu-delphi/covidcast-modeling/blob/devel/indicator-heterogeneity-I/indicator_heterogeneity_I/reports/DAP_Report_IndicatorHeterogeneity.html#predictive-ability).

Biggest change is the addition of 5-20 days ahead.  Unfortunately the results are negative:

> Disappointingly, the models with the static sensorized estimates do best, while the ones with the dynamically sensorized estimates tend to do worse than with just the raw indicator (!).
> (This is disappointing because the static sensorized estimates are “cheating”, i.e., in sample, and the dynamic sensorized ones are more “realistic” because they are computed in an online fashion). Perhaps this can be viewed in light of the fact that dynamic sensorization tends to degrade time-wise correlations – and this notion of “proper ordering across time” is important for forecasting. (Evidence of overfitting?)

I also updated the visual schema for the 7, 14 day ahead table of results so that it is easier to read (data / conclusions remain the same).

huisaddison commented:
The [predictive ability](https://htmlpreview.github.io/?https://github.com/cmu-delphi/covidcast-modeling/blob/devel/indicator-heterogeneity-I/indicator_heterogeneity_I/reports/DAP_Report_IndicatorHeterogeneity.html#predictive-ability) section has been updated.

> The results are mixed, and I’m not sure how to interpret them. Before October 15th, the models which used static sensorized features almost always performed the best. After October 15th, sometimes dynamic sensorization provides the most benefit, sometimes static sensorization. Disappointingly, dynamic sensorization always fares worse for Doctor Visits, the indicator that stands to benefit the most from sensorization.

<img width="1173" alt="Screen Shot 2020-12-18 at 14 52 45" src="https://user-images.githubusercontent.com/8422804/102668406-b1b3e000-4140-11eb-847d-3c1e810abcfb.png">

ryantibs commented:
Thanks and sorry for the short delay!  I'm going to respond explaining what I think is going on, but the main message is that I think you should just write this explanation into your DAP clearly, and we should consider it a success, rather than having you do anything more. 

1. Just a clarification actually: when you look at hospitalizations here, is the response (in the forecast problem) always deaths?  Even though the sensorizatoin target is cases in one row, and deaths in the next?

2. I think I know what's happening here.  It's exactly as you said: when we sensorize dynamically, we're losing a consistent meaning across time.  But here's how we'd fix it!  We only publish a sensorized signal in order to do things like plot a choropleth map.  But whenever we use it for modeling, we actually think of the main deliverable of your analysis (per location and time) as the **fitted sensorization model** (in our case, fitted slope and intercept).  So when we go to train a forecaster a date t, we apply this fitted sensorization model retrospectively to all data in a particular location, over the training period we're using.  That way, it always has a consistent meaning temporally, AND it's adjusted to be relevant across locations.  Make sense?

To reiterate, I'm not suggesting that you do #2, just record permanently the observations we've been making here in the DAP itself, and highlight it as an important direction for future work---whoever wants to work further developing sensorization so that it's useful for forecasting.  Then I'll close this PR and we can celebrate your DAP as complete!

ryantibs commented:
Actually, let me register one more idea which I like better than what I wrote above in point #2.  (Given how well the static sensorization has worked.). 

Consider the following sensorization scheme: at any point in time, sensorize based on all available history.  (Possibly with a buffer, i.e., at time t we use data up through t - k, if we don't believe recent data to be reasonable due to backfill as in claims signals.)

Then critically, we don't just define the most recent sensors this way, but we go back and redefine *all sensors* this way.  So at time t, we both define a sensor at time t, and redefine sensors at all times 1 through t-1, based on training a sensor model on all data available to us.  On the API side, this can be stored by appropriate use of the "as of" parameter.  A sensor is first define at time t, but is redefined "as of" time t + k, for all k > 0.  

This would be memory intensive, but a clean solution both conceptually and programmatically.  To build a forecast model at time t, we pull all sensors "as of" time t.  Due to the way we've constructed them, they'd all be comparable over time: because they'd all be outputs of the exact same model.  Make sense?

huisaddison commented:
Thanks Ryan!

1. The prediction targets are the same as the forecasting targets.  So, when we sensorized Hospitalizations against Cases, the forecasting target was also Cases, not Deaths.  (I can see why you asked -- it does make sense to evaluate both sensorized Hospitalizations against Deaths).  Would it be fruitful for us to tweak this part of the DAP? 
2. Great idea re: a global re-sensorization (using data up until k days in the past).  I revised the text to include this recommendation in the introduction.  I also revised the predictive ability section to solely discuss the 5-20 day ahead results (and moved all the original predictive ability content into the Appendix).

New text in introduction:

> **Our recommendation.**
> We find that both static and dynamic sensorization the improve spatial
> interpretation
> of the indicators, but dynamic sensorization can hurt temporal interpretation
> (and predictive ability) because the fitting of separate models in a
> moving window fashion can disrupt consistency of magnitude over time.  (Notions
> of "spatial interpretation", "temporal interpretation",
> and "predictive ability" will be formalized in the sequel).  We recommend
> a solution that attempts the "best of both worlds": on each day, we
> re-sensorize the entire past, using data up until the recent past (e.g.,
> 1 week in the past) to guard against the instability of recent data due
> to backfill.  More precisely, (for each location) on each day $t$, we would fit
> a linear model by regressing the targets $y_1, y_2, \dotsc, y_{t-k}$ on
> the indicator values $x_1, x_2, \dotsc, x_{t-k}$, and we use this model
> to produce sensorized values
> $\tilde x_1^{(t)}, \tilde x_2^{(t)}, \dotsc, \tilde x_t^{(t)}$.  Therfore,
> at each time $t$, we re-sensorize the entire indicator.  This allows us to
> partake in the spatial interpretation benefits of sensorization without
> suffering from unnecessarily degradation of the timewise interpretation.

New text in predictive ability section:

> 
> We see that when we aggregate prior to October 15th, static sensorization
> consistently does better than dynamic sensorization.  After October 15th,
> sometimes static sensorization does better, sometimes dynamic sensorization
> does better.  What gives?  Our hypothesis is that dynamic sensorization's
> ability to change the scale of the sensorized estimates over time is its
> folly.  When we separately fit models over time, we lose a consistent meaning
> temporally, which hampers our ability to perform forecasting (which seeks
> to make a statement about a relative point in time, specifically, the
> future).  This hypothesis is supported by the fact that the portions of
> time where dynamic sensorization suffers relative to static sensorization
> is precisely the portions of time where the time-wise correlations
> of the dynamic sensorized estimates are severely degraded relative to
> the those of the static sensorized estimates.

![Screenshot from 2020-12-19 23-36-43](https://user-images.githubusercontent.com/8422804/102707825-22e3b800-4253-11eb-8832-13791af08d70.png)

ryantibs commented:
Looks great, thank you.  Recommend two trivial changes then I promise I will merge :) 
1. Please put my name on as stakeholder.
2. Please remove (just comment out) the predictive experiment where we use hospitalization to predict cases.  The reversal of the causality arrow here is just too weird (in my opinion).

huisaddison commented:
[Done](https://htmlpreview.github.io/?https://github.com/cmu-delphi/covidcast-modeling/blob/devel/indicator-heterogeneity-I/indicator_heterogeneity_I/reports/DAP_Report_IndicatorHeterogeneity.html#predictive-ability) :)

ryantibs commented:
Closed!  Congratulations.  Awesome job with this DAP.
